Lineval

vicsn opened the pull request:
**Requires resampling of genesis and proving keys**

## Motivation

Add support for lineval (part of Varuna), ~10% faster proving time for a proof of transfer and simpler and a base for supporting custom gates. Some context on what was done in each round:
- Indexing: slightly updated, no more use of bivariate derivatives
- Round 1: no more commitment to `z_a` and `z_b`
- Round 2: introduced a new round just checking `z_a(X)*z_b(X)=z_c(X)`, the next rounds are all one number higher now.
- Round 3: does lineval instead of lincheck. instead of check z_M(X) = z(X)*M(X), we just prove equivalence at an evaluation `alpha`
- Round 4: updated to reflect the new indexing
- Round 5: no change.

## Test Plan
Manually checked the setup time got worse and prover time improved.

We will add Sage test vectors in future PRs.

## Related PRs

https://github.com/AleoHQ/protocol-docs/pull/17

ljedrz commented on algorithms/src/fft/polynomial/dense.rs:
```diff
@@ -224,10 +225,10 @@ impl<'a, F: Field> AddAssign<&'a DensePolynomial<F>> for DensePolynomial<F> {
-            }
+        }
+        // If the leading coefficient ends up being zero, pop it off.
+        while self.coeffs.last().unwrap().is_zero() {
```
```suggestion
        while self.coeffs.last().map(|c| c.is_zero()) == Some(true) {
```
while I'm guessing it's guaranteed that there are always elements in the collection, this safe alternative is free

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -40,21 +37,37 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
+    rand_assignments: &[F],
+    circuit_id: Option<CircuitId>,
+) {
+    // TODO: pass 3-sized array
+    let mut zk_vars = Vec::with_capacity(3);
+    for i in 0..3 {
+        zk_vars.push(cs.alloc(|| format!("random_{i}"), || Ok(rand_assignments[i])).unwrap());
+    }
```
```suggestion
    let zk_vars = (0..3).map(|i| cs.alloc(|| format!("random_{i}"), || Ok(rand_assignments[i])).unwrap());
```

vicsn replied:
See comment below

ljedrz commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -40,21 +37,37 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
+    }
+    cs.enforce(
+        || format!("constraint zk {:?}", circuit_id),
+        |lc| lc + zk_vars[0],
+        |lc| lc + zk_vars[1],
+        |lc| lc + zk_vars[2],
```
```suggestion
        |lc| lc + zk_vars.next().unwrap(),
        |lc| lc + zk_vars.next().unwrap(),
        |lc| lc + zk_vars.next().unwrap(),
```

vicsn replied:
Unfortunately: `cannot borrow as mutable because previous closure requires unique access
second borrow occurs here`. The three closures all want to capture `zk_vars`. I think the overhead of the original is minimal.

ljedrz commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -230,12 +204,30 @@ pub(crate) fn arithmetize_matrix<F: PrimeField>(
+    for row in matrix.iter() {
+        for (_, input_var_index) in row {
+            let c_i = variable_domain.reindex_by_subdomain(input_domain, *input_var_index)?;
+            c_i_to_k.entry(c_i).or_insert(vec![]).push(k);
```
```suggestion
            c_i_to_k.entry(c_i).or_default().push(k);
```

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fifth.rs:
```diff
@@ -0,0 +1,71 @@
+
+impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+    /// Output the number of oracles sent by the prover in this round.
+    pub fn num_fifth_round_oracles() -> usize {
```
```suggestion
    pub const fn num_fifth_round_oracles() -> usize {
```

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,268 @@
-    pub fn num_fourth_round_oracles() -> usize {
-        1
+    /// Output the number of oracles sent by the prover in the fourth round.
+    pub fn num_fourth_round_oracles(circuits: usize) -> usize {
```
```suggestion
    pub const fn num_fourth_round_oracles(circuits: usize) -> usize {
```

vicsn replied:
I assume this comment was made in a past version when we didn't pass arguments to the function yet.

ljedrz replied:
indeed, but this one should still be `const`-able, unless the other implementations are more advanced

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,268 @@
+                ]
+                .into_iter()
+                .map(|info| (info.label().into(), info))
+                .collect::<BTreeMap<PolynomialLabel, PolynomialInfo>>()
```
since everything gets collected into a single `BTreeMap` in the end, are these intermittent collects (incurring ordering and allocations) needed?

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,268 @@
+                );
+                (*circuit, result)
+            });
+        }
```
looks like a good dedup via fn or macro potential

vicsn replied:
Well spotted, or a simple loop! febb1da7e

ljedrz commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -672,20 +717,10 @@ where
+                            w_comm.w,
+                        )]
                     })
                     .collect_vec();
```
I'm wondering if this call could be dropped, especially since now there is only a single `LabeledCommitment` involved

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,191 @@
+                    let mut rowcheck = multiplier_2.multiply().unwrap();
+                    cfg_iter_mut!(rowcheck.coeffs).zip(&z_c.coeffs).for_each(|(ab, c)| *ab -= c);
+
+                    if !rowcheck.coeffs.iter().filter(|&c| *c != F::zero()).collect_vec().is_empty() {
```
```suggestion
                    if !rowcheck.coeffs.iter().find(|&c| *c != F::zero()).is_none() {
```
avoids an allocation

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,267 @@ use crate::{
     /// Output the number of oracles sent by the prover in the third round.
-    pub fn num_third_round_oracles(circuits: usize) -> usize {
-        circuits * 3
+    pub fn num_third_round_oracles() -> usize {
```
```suggestion
    pub const fn num_third_round_oracles() -> usize {
```

vicsn replied:
febb1da7e

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,267 @@ use crate::{
+            }
+        }
+
+        let mut matrixSums = num_instances.iter().map(|n| Vec::with_capacity(*n)).collect_vec();
```
```suggestion
        let mut matrix_sums = num_instances.iter().map(|n| Vec::with_capacity(*n)).collect_vec();
```

vicsn replied:
febb1da7e

ljedrz reviewed (commented):
Left some suggestions.

ljedrz reviewed (commented):
LGTM engineering-wise :ok_hand:.

vicsn commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -131,77 +127,57 @@ pub struct MatrixArithmetization<F: PrimeField> {
     }
 
-    batch_inversion::<F>(&mut inverses);
+    let mut row_col_indices: Vec<_> = row_indices.iter().zip_eq(&col_indices).map(|(row, col)| *row * col).collect();
```
Should rename this to `row_col_val`

Pratyush commented on algorithms/src/fft/polynomial/dense.rs:
```diff
@@ -201,7 +202,7 @@ impl<'a, 'b, F: Field> Add<&'a DensePolynomial<F>> for &'b DensePolynomial<F> {
             cfg_iter_mut!(result.coeffs).zip(&self.coeffs).for_each(|(a, b)| *a += b);
             // If the leading coefficient ends up being zero, pop it off.
-            while result.coeffs.last().unwrap().is_zero() {
+            while result.coeffs.last().map(|c| c.is_zero()) == Some(true) {
```
```suggestion
            while let Some(true) = result.coeffs.last().map(|c| c.is_zero()) {
```

vicsn replied:
https://github.com/AleoHQ/snarkVM/commit/36bc947b7ba2c575533e61586fbfc4d51c253b5b

Pratyush commented on algorithms/src/fft/polynomial/dense.rs:
```diff
@@ -224,10 +225,10 @@ impl<'a, F: Field> AddAssign<&'a DensePolynomial<F>> for DensePolynomial<F> {
-            }
+        }
+        // If the leading coefficient ends up being zero, pop it off.
+        while self.coeffs.last().map(|c| c.is_zero()) == Some(true) {
```
```suggestion
        while let Some(true) = self.coeffs.last().map(|c| c.is_zero()) {
```

vicsn replied:
36bc947b7

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -110,33 +108,22 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         let num_non_zero_b = info.num_non_zero_b;
         let num_non_zero_c = info.num_non_zero_c;
         [
             EvaluationDomain::<F>::compute_size_of_domain(num_constraints).unwrap() - 2,
```
Do we ever perform a sumcheck over the constraint domain? I don't think so, right? We only need degree bounds for domains over which we perform sumchecks.

vicsn replied:
Great catch: 214a406cd. This will likely be added again for lookups.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -174,108 +135,38 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-            _ => None,
-        }
-    }
-}
+pub struct Witness<F: PrimeField>(LabeledPolynomial<F>);
```
We can just make this a `pub type Witness<F> = ...`

vicsn replied:
080107284

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,178 @@
+        let verifier::FirstMessage { batch_combiners, .. } = verifier_message;
 
-        let summed_z_m_and_t = Self::calculate_summed_z_m_and_t(&state, *alpha, *eta_b, *eta_c, batch_combiners);
+        let h_0 = Self::calculate_lhs_second(&mut state, batch_combiners)?;
```
Should we give this a more informative name?

vicsn replied:
Absolutely, took a stab: 46106cd8c
Witness seems generic enough to be consistent across rounds and for the next lookup PR.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,178 @@
+                    let mut rowcheck = multiplier_2.multiply().unwrap();
+                    cfg_iter_mut!(rowcheck.coeffs).zip(&z_c.coeffs).for_each(|(ab, c)| *ab -= c);
+
+                    if rowcheck.coeffs.iter().any(|&c| c != F::zero()) {
+                        instance_lhs += &(&rowcheck * instance_combiner);
+                    }
```
We should be able to remove this check, no? rowcheck should never be zero unless z_a and z_b are themselves zero.

vicsn replied:
Yes, leftover from an earlier wrong version: 755a1421d

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,178 @@
+                    // Substituting in s_i, we get that lhs_i * s_i / v_R = lhs_i / v_K * (R_i.size() / R.size());
+
+                    let selector_rowcheck_time =
+                        start_timer!(|| format!("Compute selector rowcheck for {}", _circuit_id));
```
```suggestion
                        start_timer!(|| format!("Compute selector rowcheck for {_circuit_id}"));
```

vicsn replied:
8a9f588e1

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,178 @@
+            cfg_reduce!(cfg_into_iter!(job_pool.execute_all()), || Ok(DensePolynomial::zero()), |a, b| {
+                if let Ok(mut b) = b {
+                    if let Ok(mut a) = a {
+                        if b != DensePolynomial::zero() {
```
```suggestion
                        if !b.is_zero() {
```

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -15,360 +15,178 @@
+        let h_sum_time = start_timer!(|| "AHP::Prover::SecondRound h_sum");
+        let h_sum: DensePolynomial<F> =
+            cfg_reduce!(cfg_into_iter!(job_pool.execute_all()), || Ok(DensePolynomial::zero()), |a, b| {
+                if let Ok(mut b) = b {
+                    if let Ok(mut a) = a {
+                        if b != DensePolynomial::zero() {
+                            a += &std::mem::take(&mut b);
+                        }
+                        return Ok(a);
+                    }
+                }
+                Err(anyhow::anyhow!("Could not sum h instances"))
```
```suggestion
                a.and_then(|a| b.map(|b| a + &b))
              
```

vicsn replied:
That. Is. beautiful. 1da3db9e2
FYI DensePolynomial doesn't have a `+` API and I assumed this was intentioned design

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
-            })
-            .collect()
+    /// Output the degree bounds of oracles in the first round.
+    pub fn third_round_polynomial_info(domain_size: usize) -> BTreeMap<PolynomialLabel, PolynomialInfo> {
```
```suggestion
    pub fn third_round_polynomial_info(variable_domain_size: usize) -> BTreeMap<PolynomialLabel, PolynomialInfo> {
```

vicsn replied:
52082fc39

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+    /// Output the degree bounds of oracles in the first round.
+    pub fn third_round_polynomial_info(domain_size: usize) -> BTreeMap<PolynomialLabel, PolynomialInfo> {
+        [
+            PolynomialInfo::new("g_1".into(), Some(domain_size - 2), Self::zk_bound()),
```
``
```suggestion
            PolynomialInfo::new("g_1".into(), Some(variable_domain_size - 2), Self::zk_bound()),
```

vicsn replied:
52082fc39

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
-        assert!(oracles.matches_info(&Self::third_round_polynomial_info(
-            state.circuit_specific_states.keys().map(|c| (c.id, &c.index_info))
-        )));
+        let mut sumcheck_lhs = h_1.mul_by_vanishing_poly(max_variable_domain);
+        sumcheck_lhs += &x_g_1_sum;
```
Let's put this also inside the `#[cfg(debug_assertions)]`

vicsn replied:
dcb2264bb

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+
+        let oracles = prover::ThirdOracles {
+            g_1: LabeledPolynomial::new("g_1".into(), g_1, Some(max_variable_domain.size() - 2), zk_bound),
+            h_1: LabeledPolynomial::new("h_1".into(), h_1, None, None),
```
Maybe we can change the `LabeledPolynomial::new` API to `fn new(label: impl Into<String>, polynomial: DensePoly, degree_bound: impl Into<Option<usize>>, hiding_bound: impl Into<Option<usize>>)` so that we don't have to manually call `into()` and annotate with `Some`

eg the above two lines would then become:
```rust
g_1: LabeledPolynomial::new("g_1", g_1, max_variable_domain.size() - 2, zk_bound),
h_1: LabeledPolynomial::new("h_1", h_1, None, None)
```

vicsn replied:
601f8ec03

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/indexer.rs:
```diff
@@ -194,27 +191,28 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
 
-        let (constraint_domain_elements, constraint_domain_eq_poly_vals) =
-            precomputation_for_matrix_evals(&constraint_domain);
+        let constraint_domain_elements = constraint_domain.elements().collect::<Vec<_>>();
```
We can do an optimization that if `variable_domain == constraint_domain`, we can simply set `let variable_domain_elements = &constraint_domain_elements`. (This case happens for most real world circuits

vicsn replied:
Nice! b1bd1e1c3

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/indexer.rs:
```diff
@@ -194,27 +191,28 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        let variable_domain_elements = variable_domain.elements().collect::<Vec<_>>();
 
         let [a_evals, b_evals, c_evals]: [_; 3] =
             cfg_into_iter!([(&a, &non_zero_a_domain), (&b, &non_zero_b_domain), (&c, &non_zero_c_domain),])
```
```suggestion
            cfg_into_iter!([(&a, &non_zero_a_domain), (&b, &non_zero_b_domain), (&c, &non_zero_c_domain)])
```

vicsn replied:
30a4b96f4

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/indexer.rs:
```diff
@@ -194,27 +191,28 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
                 .map(|(matrix, non_zero_domain)| {
                     matrix_evals(
                         matrix,
                         non_zero_domain,
```
shouldn't we also be passing in `constraint_domain` here, as that indexes the rows?

vicsn replied:
Yeah this assymetry seemed weird to me as well. But the only reason we use the `variable_domain`, is because of how we're storing variable indexes in a locally correct format: `variable_domain.reindex_by_subdomain(input_domain, *input_var_index)`.

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -40,21 +38,38 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
-                };
-                *row_map.entry(column).or_insert_with(F::zero) += *fe;
+            row.iter().for_each(|(val, column)| {
+                if *val != F::zero() {
```
```suggestion
                if !val.is_zero() {
```

vicsn replied:
8907510b8

Just FYI, this inspired: https://github.com/AleoHQ/snarkVM/pull/1796
And depending on the response to that PR I'd prefer to `ensure!` this doesn't  happen.

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -40,21 +38,38 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
+        }
+    }
+
+    let mut zk_vars = Vec::with_capacity(3);
+    for (i, assignment) in assignments.into_iter().enumerate() {
+        zk_vars.push(cs.alloc(|| format!("random_{i}"), || Ok(assignment)).unwrap());
+    }
```
```suggestion
    let zk_vars = assignments.into_iter().enumerate().map(|(i, assignment)| {
        cs.alloc(|| format!("random_{i}"), || Ok(assignment)).unwrap());
    }).collect::<Vec<_>>();
```

vicsn replied:
c901b44f0

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -124,81 +120,67 @@ pub struct MatrixArithmetization<F: PrimeField> {
     }
 
-    batch_inversion::<F>(&mut inverses);
+    let mut row_col_indices: Vec<_> = row_indices.iter().zip_eq(&col_indices).map(|(row, col)| *row * col).collect();
```
```suggestion
    let mut row_col_indices: Vec<_> = cfg_iter!(row_indices).zip_eq(&col_indices).map(|(row, col)| *row * col).collect();
```

vicsn replied:
68a1d7131

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -124,81 +120,67 @@ pub struct MatrixArithmetization<F: PrimeField> {
     for _ in count..non_zero_domain.size() {
-        col_vec.push(elems[0]);
-        row_vec.push(elems[0]);
-        val_vec.push(F::zero());
+        row_indices.push(F::one());
+        col_indices.push(F::one());
+        row_col_indices.push(F::one());
+        row_col_vals.push(F::zero());
     }
```
Let's do the padding before we construct `row_col_indices`, as the additional pushes currently cause `row_col_indices` to reallocate

vicsn replied:
We should also avoid superfluous multiplication of row and col. Tried another approach here which first allocates the full vector, then only multiplies a (sub)slice of it: 92ec45d61

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -224,18 +206,36 @@ pub(crate) fn arithmetize_matrix<F: PrimeField>(
     }
 }
 
+pub(crate) fn non_zero_entries_per_column<F: PrimeField>(
+    matrix: &Matrix<F>,
+    variable_domain: &EvaluationDomain<F>,
+    input_domain: &EvaluationDomain<F>,
+) -> Result<BTreeMap<usize, Vec<usize>>, anyhow::Error> {
+    let mut c_i_to_k = BTreeMap::<usize, Vec<usize>>::new();
+    let mut k = 0;
+    for row in matrix.iter() {
+        for (_, input_var_index) in row {
+            let c_i = variable_domain.reindex_by_subdomain(input_domain, *input_var_index)?;
+            c_i_to_k.entry(c_i).or_default().push(k);
+            k += 1;
+        }
+    }
+    Ok(c_i_to_k)
+}
```
Can you add some documentation for what this function does? E.g. what is `k`?

vicsn replied:
bfead31c1

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/oracles.rs:
```diff
@@ -51,60 +51,59 @@ impl<F: PrimeField> FirstOracles<F> {
+        let WitnessPoly(w_poly) = self.clone();
 
-        let mut z_b_copy = LabeledPolynomialWithBasis { polynomial: vec![], info: self.z_b.info().clone() };
-        std::mem::swap(&mut self.z_b, &mut z_b_copy);
-        [w_poly.into(), z_a_copy, z_b_copy].into_iter()
+        [w_poly.into()].into_iter()
     }
 
     /// Iterate over the polynomials output by the prover in the first round.
     /// Intended for use when opening.
     pub fn iter_for_open(&self) -> impl Iterator<Item = &LabeledPolynomial<F>> {
-        [(&self.w_poly), &self.z_a_poly, &self.z_b_poly].into_iter()
+        [(&self.0)].into_iter()
     }
```
We can unify these functions now, since they return the same thing

vicsn replied:
Ah, thx 8614ed484

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+                end_timer!(z_time);
+
+                for (label, matrix_combiner, arith) in itertools::izip!(matrix_labels, matrix_combiners, matrix_arith) {
+                    let lineval_assignment = assignment.clone(); // further below, PolyMultiplier does not want just a reference
```
we can add an API for `PolyMultiplier` to take a reference too.

vicsn replied:
Ha turns out it was already there: 5551f024c

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+                    let col_at_k = arith.evals_on_K.col[k];
+                    let row_at_k = arith.evals_on_K.row[k];
+                    let row_col_val_at_k = arith.evals_on_K.row_col_val[k];
+                    let l_row_at_k = constraint_domain.size_inv * v_R.evaluate(alpha) / (alpha - row_at_k);
```
We can evaluate `v_R` at alpha once outside the map

Pratyush replied:
Ditto for the inversion: we should use batch_inversion to compute `(alpha - row_at_k).inverse()`. Otherwise this loop will be way too slow.

vicsn replied:
Great points, will be mindful of inversion in the future - c72f41937

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+
+        let v_R: SparsePolynomial<F> = constraint_domain.vanishing_polynomial();
+
+        let m_at_alpha_evals_time = start_timer!(|| format!("Compute m_at_alpha_evals parallel for {_label}"));
+        let mut m_at_alpha_evals: BTreeMap<usize, F> = cfg_iter!(c_i_to_k)
+            .map(|(&c_i, ks)| {
+                let mut m_at_c_i = F::zero();
+                for &k in ks {
+                    let col_at_k = arith.evals_on_K.col[k];
+                    let row_at_k = arith.evals_on_K.row[k];
+                    let row_col_val_at_k = arith.evals_on_K.row_col_val[k];
+                    let l_row_at_k = constraint_domain.size_inv * v_R.evaluate(alpha) / (alpha - row_at_k);
+                    // instead of calculating L_col(k), we just add val(k)*L_row(k) directly to the evaluation where we know L_col(k) will be 1
+                    // recall that row_col_val(k) = M_{i,j}*row(k)*col(k). Given that we 'ignore' L_col(k), we divide out the col(k) factor
+                    m_at_c_i += row_col_val_at_k * l_row_at_k / col_at_k;
+                }
+                (c_i, m_at_c_i)
+            })
```
We don't need to compute the evaluations of M(\alpha, c) for all c \in C manually like this. Instead, let's use the following strategy: compute the transpose of the matrix (in the offline phase? but it's fine to do it online too), and then, we can iterate as follows:
```rust
for col in matrix_transpose {
    let mut sum_for_col = F::zero();
    for (val, row_index) in col {
        sum_for_col += val * L_row(alpha)
    }
}
```
We can precalculate L_row(alpha) outside the loop too (`constraint_domain.evaluate_all_lagrange_polynomials(alpha)`)
This should be simpler to understand, and cleaner to audit.

It would also be nice to have some comments indicating what this code block is doing.

vicsn replied:
That is a lot cleaner indeed, see: a161a6d73 and 81d98173f
No need to precompute in the offline phase as for a "normal" transfer credit proof, `calc_transposes` takes less than 1% of the 3rd round's time.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -21,249 +20,286 @@ use crate::{
+        let sum = z_m_at_alpha.evaluate_over_domain_by_ref(*variable_domain).evaluations.into_iter().sum::<F>();
+        end_timer!(z_m_at_alpha_time);
+
+        // Let C = largest_domain;
+        // Let C_i = domain;
+        // Let v_C := C.vanishing_polynomial();
+        // Let v_C_i := C_i.vanishing_polynomial();
+        // Let s_i := C.selector_polynomial(C_i) = (v_C / v_C_i) * (C_i.size() / C.size());
+        // Let m := masking polynomial
+        // Let c_i := circuit combiner
+        // Let lhs_i := circuit_specific partial left hand size of the sumcheck, see protocol docs for more details
+
+        // Instead of just multiplying each lhs_i by `s_i*c_i`, we reorder the sumcheck to cancel out some terms
+        // This removes a mul and div by v_C operation over each circuit's (max_variable_domain - variable_domain)
+
+        // Later on, we divide each `lhs_i` and m by v_C, then multiply by `s_i*c_i` and sum them.
+        // Substituting in s, we get that:
+        // (\sum_i{lhs_i} + m)/v_C = \sum{h_1_i*v_C + x_g_1_i}
+        // \sum_i{c_i*s_i*(lhs_i/v_C - x_g_1_i)} + (m/v_C - x_g_1_m) = \sum{h_1_i*v_C}
+        // \sum_i{c_i*(C_i.size()/C.size())*(lhs_i/v_C_i - x_g_1_i*v_C/v_C_i)} + (m/v_C - x_g_1_m) = \sum{h_1_i*v_C}
+        // \sum_i{c_i*(C_i.size()/C.size())*(lhs_i/v_C_i} + m/v_C = \sum{h_1_i*v_C} + \sum{c_i*x_g_1_i*(v_C/v_C_i)*(C_i.size()/C.size())} + x_g_1_m
+        // (\sum_i{c_i*s_i*lhs_i} + m)/v_C = \sum{h_1_i*v_C} + \sum{c_i*s_i*x_g_1_i} + x_g_1_m
+        // (\sum_i{c_i*s_i*lhs_i} + m)/v_C = h_1*v_C + x_g_1
```
I think by now we're repeating this logic 3 times, once in round 1, once here, and then in round 5. Maybe it's worth abstracting out at this point

vicsn replied:
Good idea, they are in round 2, 3 and 4, took a stab at it: d6d5f11d3
Just FYI there are two funky differences:
- in round 3, we also want each circuit poly's remainder to add up, this is what the `remainder_witness` flag is for
- in round 4, we are not applying randomizers yet as we're getting those from the Verifier.

Pratyush commented on algorithms/src/snark/marlin/ahp/verifier/verifier.rs:
```diff
@@ -66,16 +64,16 @@ impl<TargetField: PrimeField, MM: MarlinMode> AHPForR1CS<TargetField, MM> {
                 EvaluationDomain::new(circuit_info.num_constraints).ok_or(AHPError::PolynomialDegreeTooLarge)?;
             end_timer!(constraint_domain_time);
 
+            let variable_domain_time = start_timer!(|| format!("Constructing constraint domain for {circuit_id}"));
```
Let's change the definition of `start_timer!` to something like 
```rust
macro_rules! start_timer { 
    ($e:expr) => {
        let label = format!(expr);
        // existing `start_timer` code.
        
        That way we can omit the `||` and the format from every call
    },
}
```

(The best thing to do would be to just copy the impl from https://docs.rs/anyhow/latest/anyhow/macro.anyhow.html

vicsn replied:
Will do in a separate PR in aleo-std once this PR is done

Pratyush commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -88,10 +88,12 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: MarlinMode> MarlinSNAR
 
             // Marlin only needs degree 2 random polynomials.
             let supported_hiding_bound = 1;
+            let supported_lagrange_sizes =
```
We actually don't use the Lagrange basis SRS anymore, so we can set this to empty

vicsn replied:
Done: 9022329a3
As I note in the TODO: shall I make a future PR removing it from the `CommitterKey`?

Pratyush commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -172,10 +174,10 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: MarlinMode> MarlinSNAR
     }
 
-    fn absorb_with_msg(commitments: &[Commitment<E>], msg: &prover::ThirdMessage<E::Fr>, sponge: &mut FS) {
+    fn absorb_with_sums(commitments: &[Commitment<E>], sums: &[prover::MatrixSums<E::Fr>], sponge: &mut FS) {
```
Why the change here?

vicsn replied:
Because now we're not just sending sums in the matrix_sumcheck (now the fourth round), but also the lineval_sumcheck (now the third round). So I'm calling `absorb_labeled_with_sums` after both rounds.

Pratyush commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -423,8 +428,12 @@ where
-        let (prover_third_message, third_oracles, prover_state) =
-            AHPForR1CS::<_, MM>::prover_third_round(&verifier_second_msg, prover_state, zk_rng)?;
+        let (prover_third_message, third_oracles, prover_state) = AHPForR1CS::<_, MM>::prover_third_round(
+            &verifier_first_message,
```
We should just store `verifier_first_message` in `prover state`, so that we don't have to pass it explicitly.

Also, it seems worth while to make `prover_xyz_round` methods into methods on `ProverState`, so that we can call them like

`let (prover_third_message, third_oracles) = prover_state.third_round(...)`

(The round functions mutate `state` as they go along).

Similarly for `VerifierState`.

vicsn replied:
Good idea, will look a lot cleaner. I will also think about how the AHP and State objects should cleanly relate. I started on this but this will be another massive diff and doesn't require any cryptography review, so I suggest I fix this in a future PR.

vicsn replied:
https://github.com/AleoHQ/snarkVM/pull/1875

Pratyush commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -469,41 +502,42 @@ where
             .collect();
         assert!(
             polynomials.len()
                 == keys_to_constraints.len() * 12 + // polys for row, col, rowcol, val
```
Let's make a variable `num_unique_circuits = keys_to_constraints.len()`, and use that for these assertions. Slightly clearer IMO

vicsn replied:
00c681431

Yes, in a future PR I'll also make a `num_index_oracles` function for consistency

Pratyush commented on algorithms/src/snark/marlin/marlin.rs:
```diff
@@ -667,27 +717,16 @@ where
         let mut first_comms_consumed = 0;
         let mut first_commitments = batch_sizes
             .iter()
-            .flat_map(|(&circuit_id, &batch_size)| {
+            .flat_map(|(circuit_id, &batch_size)| {
                 let first_comms = comms.witness_commitments[first_comms_consumed..][..batch_size]
                     .iter()
                     .enumerate()
                     .flat_map(|(j, w_comm)| {
-                        [
-                            LabeledCommitment::new_with_info(
-                                &first_round_info[&witness_label(circuit_id, "w", j)],
-                                w_comm.w,
-                            ),
-                            LabeledCommitment::new_with_info(
-                                &first_round_info[&witness_label(circuit_id, "z_a", j)],
-                                w_comm.z_a,
-                            ),
-                            LabeledCommitment::new_with_info(
-                                &first_round_info[&witness_label(circuit_id, "z_b", j)],
-                                w_comm.z_b,
-                            ),
-                        ]
-                    })
-                    .collect_vec();
+                        [LabeledCommitment::new_with_info(
+                            &first_round_info[&witness_label(*circuit_id, "w", j)],
+                            w_comm.w,
+                        )]
+                    });
```
We can rewrite this as a `map`:

```suggestion
                   .map(|(j, w_comm)| LabeledCommitment::new_with_info(
                        &first_round_info[&witness_label(*circuit_id, "w", j)],
                        w_comm.w,
                    ));

vicsn replied:
Good catch: 2c5a44a34

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -160,14 +160,14 @@ pub(crate) fn matrix_evals<F: PrimeField>(
-    cfg_iter_mut!(row_col_indices[0..non_zero_entries]).zip_eq(col_par_iter).for_each(|(rc, &col)| *rc *= col);
-    let row_col_par_iter = cfg_iter!(row_col_indices[0..non_zero_entries]);
-    cfg_iter_mut!(row_col_vals[0..non_zero_entries]).zip_eq(row_col_par_iter).for_each(|(v, rc)| *v *= rc);
+    let col_par_iter = cfg_iter!(col_indices).take(non_zero_entries);
+    cfg_iter_mut!(row_col_indices).take(non_zero_entries).zip_eq(col_par_iter).for_each(|(rc, &col)| *rc *= col);
+    let row_col_par_iter = cfg_iter!(row_col_indices).take(non_zero_entries);
+    cfg_iter_mut!(row_col_vals).take(non_zero_entries).zip_eq(row_col_par_iter).for_each(|(v, rc)| *v *= rc);
```
I think we can just use `zip` here, and it'll truncate `col_indices` and `row_col` automatically. So that way you don't have to construct the intermediate iterators manually.

vicsn replied:
Indeed, using zip now. However, still have to make the extra iterators due to `for_each` only working when both zipped iters are made using `cfg_iter!`. 3ceb3463f

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -345,14 +349,14 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        // let (h_1_i, mut xg_1_i) = z_m_at_alpha.divide_by_vanishing_poly(*variable_domain).unwrap();
+        // xg_1_i = xg_1_i.mul_by_vanishing_poly(*max_variable_domain);
+        // let (xg_1_i, remainder) = xg_1_i.divide_by_vanishing_poly(*variable_domain).unwrap();
+        // assert!(remainder.is_zero());
```
We can delete these commented out lines

vicsn replied:
Sloppy, removed - 17cab955c

Pratyush commented on algorithms/src/polycommit/sonic_pc/mod.rs:
```diff
@@ -173,11 +176,11 @@ impl<E: PairingEngine, S: AlgebraicSponge<E::Fq, 2>> SonicKZG10<E, S> {
         ck: &CommitterUnionKey<E>,
         polynomials: impl IntoIterator<Item = LabeledPolynomialWithBasis<'b, E::Fr>>,
-        rng: Option<&mut dyn RngCore>,
+        rng: Option<&mut R>,
```
Why this change?

vicsn replied:
Hmm it did not compile during an intermediary test and did not seem to me to worsen the interface, now it does compile, reverted: e791de587

howardwu replied:
Just for some context to understand the situation:

It's better not to use dynamic dispatch when possible. The perf hit can be significant depending on context and frequency of use. I think this case would be innocent, but nonetheless not ideal.

Pratyush replied:
I think it's fine, because we only sample a couple of field elements from the rng; it's better than the alternative because when rng is `None`, it doesn't require us to specify the type of the (unused) `Rng`

howardwu commented on parameters/src/testnet3/resources/transfer.metadata:
```diff
@@ -0,0 +1,6 @@
+{
```
This file should be removed

vicsn committed:
Implement lineval sumcheck of Varuna protocol

    - update indexing
    - add lagrange polynomials to prover state
    - remove z_a and z_b from first round commitments
    - add a round for rowcheck zerocheck
    - replace lincheck by lineval sumcheck in the next round
    - temporarily remove mask_poly
    - make cargo fmt and clippy happy
    - add randomizing constraints to make the scheme ZK
    - MARLIN-2019 -> VARUNA-2023
    - We only need degree bounds for domains over which we perform sumchecks.
    - Improve naming witness gen functions
    - Make optimized selector calculation DRY
    - Remove supported_lagrange_sizes

vicsn committed:
When MM::ZK:
- Do not pass zk_rng to commit
- Do not call calculate_mask_poly

howardwu committed:
Update usage of Result

vicsn committed:
Remove unused Clone interface

vicsn committed:
Nit: use let Some(true)

vicsn committed:
Nit: complete comment

vicsn committed:
Add comment

vicsn committed:
Add assert to check correctness of fourth round

vicsn committed:
Update expected tx sizes

ljedrz commented:
@vicsn let me know (re-request review) when the conflicts are resolved and the PR is ready for another look :+1:.

edit: well that was quick :smile:

vicsn committed:
Revert paramater file changes

ljedrz commented on algorithms/src/fft/domain.rs:
```diff
@@ -314,18 +316,18 @@ impl<F: FftField> EvaluationDomain<F> {
+    /// Given an index in the `other` subdomain, return an index into this domain `self`
+    /// This assumes the `other`'s elements are also `self`'s first elements
+    pub fn reindex_by_subdomain(&self, other: &Self, index: usize) -> Result<usize> {
+        ensure!(self.size() >= other.size(), "self.size() must be smaller than other.size()");
```
```suggestion
        ensure!(self.size() >= other.size(), "other.size() must be smaller than self.size()");
```

vicsn replied:
Good catch ab096deab

ljedrz commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+                    lineval.add(sum_b_fourth * eta_b * v_X_at_beta[id], w_j.clone());
+
+                    lineval.add(sum_c_fourth * eta_c * x_at_betas[id][j], LCTerm::One);
+                    lineval.add(sum_c_fourth * eta_c * v_X_at_beta[id], w_j.clone());
```
```suggestion
                    lineval.add(sum_c_fourth * eta_c * v_X_at_beta[id], w_j);
```
this last `clone` should be avoidable

vicsn replied:
d1347c462

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,229 @@
+                .into_iter()
+                .map(|info| (info.label().into(), info))
+            })
+            .collect::<BTreeMap<PolynomialLabel, PolynomialInfo>>()
```
```suggestion
            .collect()
```

vicsn replied:
23aa8acfe

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/mod.rs:
```diff
@@ -124,6 +163,61 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            assert!(remainder.is_zero());
+            let multiplier = combiner * src_domain.size_as_field_element * target_domain.size_inv;
+            cfg_iter_mut!(h_i.coeffs).for_each(|c| *c *= multiplier);
+            end_timer!(selector_time);
```
nit: this call could be done just once after the `if-else`

ljedrz replied:
ah, nvm, the starting text differs

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
-        let oracles = prover::ThirdOracles { gs };
+        let g_1 = DensePolynomial::from_coefficients_slice(&x_g_1_sum.coeffs[1..]);
+
+        drop(x_g_1_sum);
```
a comment with the reason for this `drop` would be useful

vicsn replied:
For context, this was already there and just as secret/sensitive as a lot of other values we don't `drop`... So I guess it is defense in depth and clarity for the dev: ab096deab

Pratyush replied:
It's there to reduce the memory usage: we can drop things once they're no longer used in the rest of the algorithm.

ljedrz reviewed (commented):
Went over the PR again, left a few comments.

vicsn committed:
Improve comments

vicsn committed:
Remove Clone

vicsn committed:
Remove superfluous type hint

vicsn committed:
Make num_fourth_round_oracles const

ljedrz commented:
I'm not savvy with all the crypto involved so I can't approve this but, with the recent changes, LGTM engineering-wise :+1:.

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -206,64 +195,73 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        let v_R_at_alpha = max_constraint_domain.evaluate_vanishing_polynomial(alpha);
+        end_timer!(v_R_at_alpha_time);
+
+        #[rustfmt::skip]
```
I think we can avoid the 
```suggestion
```

vicsn replied:
276af0f90

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -206,64 +195,73 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-            );
-        }
-        end_timer!(bivariate_poly_time);
+        let v_C_at_beta_time = start_timer!(|| "v_C_at_beta");
+        let v_C_at_beta = max_variable_domain.evaluate_vanishing_polynomial(beta);
+        end_timer!(v_C_at_beta_time);
```
I think we can remove this timer; the evaluation time should be close to zero.
```suggestion
        let v_C_at_beta = max_variable_domain.evaluate_vanishing_polynomial(beta);
```

vicsn replied:
47240504d

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -206,64 +195,73 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-        end_timer!(v_H_at_beta_time);
+        let v_K_at_gamma_time = start_timer!(|| "v_K_at_gamma");
+        let v_K_at_gamma = max_non_zero_domain.evaluate_vanishing_polynomial(gamma);
+        end_timer!(v_K_at_gamma_time);
```
```suggestion
        let v_K_at_gamma = max_non_zero_domain.evaluate_vanishing_polynomial(gamma);
```

vicsn replied:
47240504d

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
             .iter()
             .enumerate()
             .map(|(i, (circuit_id, circuit_state))| {
                 let lag_at_beta = circuit_state.input_domain.evaluate_all_lagrange_coefficients(beta);
```
I think we can construct this once per `circuit_id`, instead of once per circuit_state

vicsn replied:
Hmm not sure what you mean, `circuit_specific_states` contains `(circuit_id, circuit_state)`, it is already creating only one per `circuit_id`.

For reference, within `circuit_state`, we have a `batch_size` of how many instances of that circuit we prove

Pratyush replied:
You're right, I misread the code.

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-                let constraint_domain = state.circuit_specific_states[id].constraint_domain;
-                let selector = selector_evals(&mut selectors, &largest_constraint_domain, &constraint_domain, beta);
+                let variable_domain = circuit_state.variable_domain;
+                let selector = selector_evals(&mut selectors, &max_variable_domain, &variable_domain, beta);
```
This performs one inversion per circuit; it might be worthwhile to consider a `batch_selector_evals` method which performs only a single inversion for the entire batch.

vicsn replied:
Nice idea, I think too much for this PR and as this will be easy to review by others - noted it for future work

Pratyush replied:
Fair, it's not a publicly visible change anyway.

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            let RC = circuit_state.constraint_domain.size_as_field_element
+                * circuit_state.variable_domain.size_as_field_element;
 
             let s_a = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_a_domain, gamma);
             let s_b = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_b_domain, gamma);
             let s_c = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_c_domain, gamma);
```
Ditto for this: probably worth precomputing the selector evaluations via a batch inversion method.

vicsn replied:
See comment above

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -40,21 +40,39 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
+        }
+    }
+
+    let zk_vars = assignments
+        .into_iter()
+        .enumerate()
+        .map(|(i, assignment)| cs.alloc(|| format!("random_{i}"), || Ok(assignment)).unwrap())
+        .collect_vec();
+    cs.enforce(|| "constraint zk", |lc| lc + zk_vars[0], |lc| lc + zk_vars[1], |lc| lc + zk_vars[2]);
```
if MM::ZK is false, we should just not add new variables or the new constraint.

vicsn replied:
1639d360f and 9071a88e7

Now if we don't have zero-knowledge, we hit my previously mentioned edge case: the degree ánd value of `h_0` is `0` when `num_constraints == constraint_domain.size()`

This happens because the only remaining unevaluated polynomial in the rowcheck equation is `L_row(k)^R(X)`, which becomes the degree 0 polynomial of `1` in this case because we use every row exactly once.

Do you think this might be a problem? We still only send a commitment of `h_0`, we send only a single evaluation proof, so AFAICT the degree and value of `h_0` don't leak.

The only ways I see this edge case can be avoided:
- if we construct a dummy constraint if every row is used exactly once in one or all matrices. The downside is that this doubles the constraint_domain.size()
- if we add and commit to a second masking polynomial which sums to 0 over the constraint domain. Based on Michel's writing, we'll likely need this anyway when we add and use lookups.

Pratyush replied:
I think it's fine. We're unlikely to hit that edge case in practice, but maybe we should add a test for it.

vicsn replied:
Cool. It is indeed currently "triggered" by one of our existing test cases. Thank you for the quick response.

Pratyush commented on algorithms/src/snark/marlin/ahp/matrices.rs:
```diff
@@ -124,81 +123,72 @@ pub struct MatrixArithmetization<F: PrimeField> {
+    let non_zero_entries = row_indices.len();
+
+    // Zip safety: we intentionally only multiply the first non_zero_entries
+    let col_indices_iter = cfg_iter!(col_indices);
+    cfg_iter_mut!(row_col_indices).zip(col_indices_iter).for_each(|(rc, &col)| *rc *= col);
+    let row_col_indices_iter = cfg_iter!(row_col_indices);
+    cfg_iter_mut!(row_col_vals).zip(row_col_indices_iter).for_each(|(v, rc)| *v *= rc);
```
Does this not work?
```suggestion
    cfg_iter_mut!(row_col_indices).zip(&col_indices).for_each(|(rc, &col)| *rc *= col);
    cfg_iter_mut!(row_col_vals).zip(&row_col_indices).for_each(|(v, rc)| *v *= rc);
```

vicsn replied:
Yes. d1f791aee

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/message.rs:
```diff
@@ -25,11 +28,46 @@ pub struct MatrixSums<F: PrimeField> {
+                    combiners.circuit_combiner,
+                    circuit_sums
+                        .iter()
+                        .map(|instance_sums| {
+                            instance_sums.sum_a + eta_b * instance_sums.sum_b + eta_c * instance_sums.sum_c
+                        })
```
```suggestion
                        .map(|sums| sums.sum_a + eta_b * sums.sum_b + eta_c * sums.sum_c)
```

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/message.rs:
```diff
@@ -25,11 +28,46 @@ pub struct MatrixSums<F: PrimeField> {
+            .iter()
+            .zip(batch_combiners.values())
+            .map(|(circuit_sums, combiners)| {
+                (
+                    combiners.circuit_combiner,
```
```suggestion
                    combiners.circuit_combiner *
```

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/message.rs:
```diff
@@ -25,11 +28,46 @@ pub struct MatrixSums<F: PrimeField> {
+                        .zip(combiners.instance_combiners.iter())
+                        .map(|(instance_sum, instance_combiner)| instance_sum * instance_combiner)
+                        .sum::<F>(),
+                )
+            })
+            .map(|(circuit_combiner, circuit_sum)| circuit_combiner * circuit_sum)
```
```suggestion
            })
```

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/message.rs:
```diff
@@ -25,11 +28,46 @@ pub struct MatrixSums<F: PrimeField> {
+        self.sums
+            .iter()
+            .zip(batch_combiners.values())
+            .map(|(circuit_sums, combiners)| {
+                (
+                    combiners.circuit_combiner,
+                    circuit_sums
+                        .iter()
+                        .map(|instance_sums| {
+                            instance_sums.sum_a + eta_b * instance_sums.sum_b + eta_c * instance_sums.sum_c
+                        })
+                        .zip(combiners.instance_combiners.iter())
+                        .map(|(instance_sum, instance_combiner)| instance_sum * instance_combiner)
+                        .sum::<F>(),
+                )
+            })
+            .map(|(circuit_combiner, circuit_sum)| circuit_combiner * circuit_sum)
+            .sum()
```
```suggestion
            .map(|(circuit_sums, combiners)| {
                    combiners.circuit_combiner *
                    circuit_sums
                        .iter()       
                        .zip(&combiners.instance_combiners)
                        .map(|(sums, combiner)| (sums.sum_a + eta_b * sums.sum_b + eta_c * sums.sum_c) * combiner)
                        .sum::<F>(),
            })
            .sum()
```

vicsn replied:
a722267ae

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -139,143 +100,65 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        LabeledPolynomial::new("mask_poly".to_string(), mask_poly, None, None)
     }
 
     fn calc_w(
```
In line with `calculate_mask_poly`
```suggestion
    fn calculate_w(
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,229 @@
-    pub fn fourth_round_polynomial_info() -> BTreeMap<PolynomialLabel, PolynomialInfo> {
-        [("h_2".into(), PolynomialInfo::new("h_2".into(), None, None))].into()
+    #[allow(clippy::too_many_arguments)]
+    fn calc_matrix_sumcheck_witness(
```
Consistent with `calculate_mask_poly`.

```suggestion
    fn calculate_matrix_sumcheck_witness(
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs:
```diff
@@ -12,58 +12,229 @@
+            let b_poly = {
+                let evals: Vec<F> = cfg_iter!(row_on_K.evaluations)
+                    .zip_eq(&col_on_K.evaluations)
+                    .zip_eq(&row_col_on_K.evaluations)
```
I think we should omit `row_col_on_K.evaluations` from the proving key, as it can be obtained by computing `r_c = r * c`. This should help reduce the proving key size nicely. (We should still keep `row_col` in the verifying key, of course.

vicsn replied:
Nice 2bdf9248e

Depending on our existing metrics infra, In the future I will add tests to check the pk and vk size to avoid regressions.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
-            end_timer!(a_poly_time);
-            a_poly
-        });
+    fn calc_lineval_sumcheck_witness(
```
```suggestion
    fn calculate_lineval_sumcheck_witness(
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
+        Ok((h_1_sum, xg_1_sum, msg))
+    }
+
+    fn calc_assignments(state: &mut prover::State<F, MM>) -> Result<BTreeMap<CircuitId, Vec<DensePolynomial<F>>>> {
```
```suggestion
    fn calculate_assignments(state: &mut prover::State<F, MM>) -> Result<BTreeMap<CircuitId, Vec<DensePolynomial<F>>>> {
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
+        Ok(assignments)
+    }
+
+    fn calc_matrix_transpose(
```
```suggestion
    fn calculate_matrix_transpose(
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
-            .zip_eq(&col_on_K.evaluations)
-            .map(|(r, c)| (beta - r) * (alpha - c))
+    #[allow(clippy::too_many_arguments)]
+    fn calc_lineval_sumcheck_instance_witness(
```
```suggestion
    fn calculate_lineval_sumcheck_instance_witness(
```

vicsn replied:
955fe6d13

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -12,258 +12,321 @@
+        let l_at_alpha = constraint_domain.evaluate_all_lagrange_coefficients(alpha);
+        let m_at_alpha_evals: Vec<_> = cfg_iter!(matrix_transpose)
+            .map(|col| {
+                let mut sum_for_col = F::zero();
+                for (val, row_index) in col {
+                    sum_for_col += *val * l_at_alpha[*row_index]
+                }
+                sum_for_col
```
```suggestion
				col.iter().map(|(val, row_index)| *val * l_at_alpha[*row_index]).sum::<F>()
```

vicsn replied:
826372904

vicsn committed:
Remove #[rustfmt::skip]

Pratyush commented on synthesizer/src/vm/execute.rs:
```diff
@@ -187,13 +187,13 @@ mod tests {
         // Assert the size of the transaction.
         let transaction_size_in_bytes = transaction.to_bytes_le().unwrap().len();
-        assert_eq!(1387, transaction_size_in_bytes, "Update me if serialization has changed");
+        assert_eq!(1468, transaction_size_in_bytes, "Update me if serialization has changed");
```
Why has the proof size gone up by this much? At most it should be up by 16 bytes (we removed one commitment (+ `h_0`, - `z_a` and `z_b`), and added two field elements (- z_b, + sum_a, sum_b, sum_c)

vicsn committed:
Remove tiny timers

Pratyush reviewed (approved):
LGTM modulo the foregoing changes. If tests pass after this we should be good to go.

vicsn committed:
Only add randomizing vars when MM::ZK

vicsn committed:
Simplify indexing

vicsn committed:
simplify ThirdMessage::sum

vicsn committed:
calc_ > calculate_

vicsn committed:
simplify lineval_sumcheck

vicsn committed:
Reduce proof size by not serializing Vec len

vicsn committed:
Allow h_0.degree() of 0

vicsn committed:
Remove row_col evals from proving key

vicsn committed:
cargo fmt

vicsn committed:
Clippy and fmt

vicsn committed:
Correct third_msg serialization

howardwu committed:
Merge branch 'testnet3' into lineval

howardwu committed:
Merge branch 'testnet3' into lineval

vicsn committed:
Merge remote-tracking branch 'origin/testnet3' into lineval

vicsn committed:
Update tx sizes

howardwu committed:
Merge branch 'testnet3' into lineval

howardwu committed:
Update parameters

howardwu committed:
Update block:

howardwu committed:
Regenerate expectations
